File: lib/rice_cooker_service.rb
class RiceCookerService
  attr_accessor :is_plugged_in, :is_empty, :contains_water

  def initialize(is_plugged_in, is_empty, contains_water)
    @is_plugged_in = is_plugged_in
    @is_empty = is_empty
    @contains_water = contains_water
  end

  def pour_water
    if !@contains_water
      @contains_water = true
      @is_empty = false
      'You poured water.'
    else
      "There's already water."
    end
  end

  def put_ingredient
    @is_empty = false
    'You put ingredient in.'
  end

  def check_if_cooked_without_water
    !@is_empty && !@contains_water
  end

  def plug_in
    if !@is_plugged_in
      @is_plugged_in = true
      'You plugged in. The rice cooker is ready to use.'
    else
      "It's already plugged in."
    end
  end

  def start
    if !@is_plugged_in
      'The rice cooker is not plugged in yet.'
    else
      if @is_empty
        'There\'s nothing to cook yet in the rice cooker.'
      elsif check_if_cooked_without_water
        'You put ingredient without water.'
      else
        "Ok\nYou can eat now."
      end
    end
  end

  def unplug
    @is_plugged_in = false
    'You unplugged the rice cooker.'
  end
end

File: pkg/service/rice_cooker_service.go
package service

type RiceCookerService struct {
	IsPluggedIn   bool
	IsEmpty       bool
	ContainsWater bool
}

func (rcs *RiceCookerService) PourWater() string {
	if !rcs.ContainsWater {
		rcs.ContainsWater = true
		rcs.IsEmpty = false
		return "You poured water."
	} else {
		return "There's already water."
	}
}

func (rcs *RiceCookerService) PutIngredient() string {
	rcs.IsEmpty = false
	return "You put ingredient in."
}

func (rcs *RiceCookerService) CheckIfCookedWithoutWater() bool {
	return !rcs.IsEmpty && !rcs.ContainsWater
}

func (rcs *RiceCookerService) PlugIn() string {
	if !rcs.IsPluggedIn {
		rcs.IsPluggedIn = true
		return "You plugged in. The rice cooker is ready to use."
	} else {
		return "It's already plugged in."
	}
}

func (rcs *RiceCookerService) Start() string {
	if !rcs.IsPluggedIn {
		return "The rice cooker is not plugged in yet."
	} else {
		if rcs.IsEmpty {
			return "There's nothing to cook yet in the rice cooker."
		} else if rcs.CheckIfCookedWithoutWater() {
			return "You put ingredient without water."
		} else {
			return "Ok \n" + "You can eat now."
		}
	}
}

func (rcs *RiceCookerService) UnPlug() string {
	rcs.IsPluggedIn = false
	return "You unplugged the rice cooker."
}
